File: apps/mobile/src/auth/capability.ts
/**
 * The offline capability token, on the device side.
 *
 * Mirrors `core_api.domain.auth.capability`. Seventy-two hours, one GN division, one
 * permission: drafting damage assessments. It cannot approve an entitlement, release
 * funds, read another division or read the incident queue. If the handset is lost, what
 * the finder gains is the ability to write drafts that a DS officer reviews before any of
 * them turns into money.
 *
 * The numbers below are asserted against the server's rather than imported, because the
 * two cannot share a module and a silent drift would show up as a token the server
 * refuses at the worst possible moment - three days into a cut-off division.
 */

/** 72 hours, from `core_api.domain.auth.capability.CAPABILITY_TTL`. */
export const CAPABILITY_TTL_MS = 72 * 60 * 60 * 1000;

/** The entire authority of the token. Anything added is something a lost handset can do. */
export const CAPABILITY_SCOPES: readonly string[] = ['assessment:write'];

/**
 * Refresh this far before expiry.
 *
 * Six hours, not the thirty seconds the online token uses. An officer who will be out of
 * coverage for a day needs the token renewed on the last connection they had, not on the
 * one they were going to have.
 */
export const CAPABILITY_REFRESH_SKEW_MS = 6 * 60 * 60 * 1000;

export interface CapabilityToken {
  readonly token: string;
  readonly gnDivisionCode: string;
  readonly deviceId: string;
  readonly permits: readonly string[];
  /** Epoch milliseconds. */
  readonly expiresAt: number;
}

export type CapabilityStatus =
  | { readonly usable: true; readonly renewSoon: boolean; readonly expiresInMs: number }
  | { readonly usable: false; readonly reason: 'expired' | 'wrong-division' | 'absent' };

/**
 * Whether a token may be used to draft an assessment here, now.
 *
 * `division` is the division the form is being filled in for. A token pinned to one
 * division cannot write in another, and the app checks that before the officer types
 * rather than after the sync fails - a rejection three days later is a rejection of work
 * that cannot be redone.
 */
export function capabilityStatus(
  token: CapabilityToken | null,
  { division, now }: { division: string; now: number },
): CapabilityStatus {
  if (!token) return { usable: false, reason: 'absent' };
  if (token.gnDivisionCode !== division) return { usable: false, reason: 'wrong-division' };
  if (token.expiresAt <= now) return { usable: false, reason: 'expired' };

  return {
    usable: true,
    renewSoon: token.expiresAt - now <= CAPABILITY_REFRESH_SKEW_MS,
    expiresInMs: token.expiresAt - now,
  };
}

/** Whether the token authorises this permission at all. */
export function permits(token: CapabilityToken, scope: string): boolean {
  return token.permits.includes(scope);
}

/** The server's response to `POST /api/v1/auth/capability-token`. */
export interface CapabilityTokenResponse {
  readonly capability_token: string;
  readonly expires_in: number;
  readonly gn_division_code: string;
  readonly permits: readonly string[];
}

export function fromResponse(
  response: CapabilityTokenResponse,
  { deviceId, now }: { deviceId: string; now: number },
): CapabilityToken {
  return {
    token: response.capability_token,
    gnDivisionCode: response.gn_division_code,
    deviceId,
    permits: response.permits,
    expiresAt: now + response.expires_in * 1000,
  };
}

File: services/core-api/src/core_api/domain/auth/capability.py
"""The offline capability token carried by the Field Companion.

Connectivity is the failure mode this platform exists to survive. A GN officer in a cut-off
division must be able to work for three days on nothing. The security answer to that is not
to make the app online-only - it is to make the offline credential capable of almost
nothing.

This token authorises exactly one thing: creating draft damage assessments, in one GN
division. It cannot approve an entitlement, cannot release funds, cannot read another
division, cannot read the incident queue. If the device is lost, what the finder gains is
the ability to write drafts that a DS officer will review before any of them turns into
money.

The app holds it in the platform secure enclave or Keystore. On reconnect the drafts sync
and the token is exchanged for a fresh one.
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timedelta
from typing import Final

from sarana_shared.auth.grants import ScopeGrant, ScopeType
from sarana_shared.auth.scopes import Role, Scope
from sarana_shared.domain.admin import AdminLevel, level_of
from sarana_shared.domain.time import utc_now

# Three days. Long enough to cover a division cut off by a cyclone, short enough that a
# device stolen at the start of a response is useless before the response ends.
CAPABILITY_TTL: Final = timedelta(hours=72)

# The entire authority of this token. Deliberately a single permission: anything added
# here is something a lost, unlocked field device can do.
CAPABILITY_SCOPES: Final[frozenset[Scope]] = frozenset({Scope.ASSESSMENT_WRITE})


class NotAFieldOfficer(ValueError):
    """Only a GN officer, pinned to one GN division, may hold a capability token."""


@dataclass(frozen=True, slots=True)
class CapabilityRequest:
    """A request to mint an offline token for one officer in one division."""

    subject_id: str
    gn_division_code: str
    device_id: str

    def __post_init__(self) -> None:
        if level_of(self.gn_division_code) is not AdminLevel.GN_DIVISION:
            raise NotAFieldOfficer(
                f"a capability token is pinned to a single GN division; "
                f"{self.gn_division_code!r} is not one"
            )
        if not self.device_id.strip():
            raise NotAFieldOfficer(
                "a capability token must be bound to a device, so a lost handset can be "
                "revoked without revoking the officer"
            )


def capability_grants(gn_division_code: str) -> frozenset[ScopeGrant]:
    """The grants a capability token carries: draft assessments, one division, nothing else."""
    return frozenset(
        ScopeGrant(scope=scope, scope_type=ScopeType.GN, scope_code=gn_division_code)
        for scope in CAPABILITY_SCOPES
    )


def may_hold_capability(roles: frozenset[Role]) -> bool:
    """Whether these roles permit an offline token at all.

    Only GN officers. An approver working offline would mean approvals accumulating on a
    device with no way to check them against the ledger, which is precisely the situation
    the two human gates exist to prevent.
    """
    return Role.GN_OFFICER in roles


def expires_at(*, issued_at: datetime | None = None) -> datetime:
    """When a capability token minted now stops being accepted."""
    return (issued_at or utc_now()) + CAPABILITY_TTL


def is_permitted_action(scope: Scope) -> bool:
    """Whether a capability token authorises this permission.

    Called by the authorisation dependency before any area check. A capability token is
    refused on every endpoint except assessment drafting, and the refusal says so rather
    than reporting a missing area.
    """
    return scope in CAPABILITY_SCOPES
